capsules/src/net/thread/tlv.rs

```diff
@@ -0,0 +1,1315 @@
+//! assumed to have a maximum length of 128 bytes. This assumption is made
+//! only when a variable-length value must be decoded from network byte
+//! order before it can be interpreted correctly. Excluded from this case
+//! are variable-length values that contain data that must later be decoded
+//! by the caller before being interpreted (for example, sub-TLVs). These
+//! values are instead returned as a slice of the original buffer passed to
+//! the decode function.
+//!
+//! This module currrently implements the minimum subset of TLVs required
+//! to connect a Sleepy End Device (SED) to a Thread network.
+//!
+//! Author: Mateo Garcia
+//!         mateog@stanford.edu
+
+
+// NOTES FOR DEBUGGING:
+// - .to_be() may not have been called on values wider than one byte
+// - encode_bytes_be may have been used instead of encode_bytes
+// - decode_bytes_be may have been used instead of decode_bytes
+// - See 4.5.25 Active Operational Dataset TLV and 4.5.26 Pending Operational Dataset TLV
+//    - Are Active and Pending Timestamp TLVs, respectively, required to be sent as well
+//      if either of the dataset tlvs are sent?
+
+
+use core::mem;
+use net::stream::{decode_u8, decode_u16, decode_u32, decode_bytes_be};
+use net::stream::{encode_u8, encode_u16, encode_u32, encode_bytes, encode_bytes_be};
+use net::stream::SResult;
+
+const TL_WIDTH: usize = 2; // Type and length fields of TLV are each one byte.
+const MAX_VALUE_LENGTH: usize = 128; // Assume a TLV value will be no longer than 128 bytes.
+
+pub enum Tlv<'a> {
+    SourceAddress(u16),
+    Mode(u8),
+    Timeout(u32),
+    Challenge([u8; 8]), // Byte string max length 8 bytes.
+    Response([u8; 8]), // Byte string max length 8 bytes.
+    LinkLayerFrameCounter(u32),
+    // LinkQuality,                  // TLV type Not used in Thread
+    // NetworkParameter,             // TLV type Not used in Thread
+    MleFrameCounter(u32),
+    /*
+    TODO: Not required to implement MLE for SED
+    Route64,
+    */
+    Address16(u16),
+    LeaderData {
+        partition_id: u32,
+        weighting: u8,
+        data_version: u8,
+        stable_data_version: u8,
+        leader_router_id: u8,
+    },
+    NetworkData(&'a [u8]),
+    TlvRequest(&'a [u8]),
+    ScanMask(u8),
+    Connectivity {
+        parent_priority: u8,
+        link_quality_3: u8,
+        link_quality_2: u8,
+        link_quality_1: u8,
+        leader_cost: u8,
+        id_sequence: u8,
+        active_routers: u8,
+        sed_buffer_size: Option<u16>,
+        sed_datagram_count: Option<u8>,
+    },
+    LinkMargin(u8),
+    Status(u8),
+    Version(u16),
+    /*
+    TODO: Not required to implement MLE for SED
+    AddressRegistration
+    AddressRegistration
+    Channel
+    PanId
+    ActiveTimestamp
+    PendingTimestamp
+    ThreadDiscovery
+    */
+    ActiveOperationalDataset(&'a [u8]),
+    PendingOperationalDataset(&'a [u8]),
+}
+
+impl<'a> Tlv<'a> {
+    pub fn encode(&self, buf: &mut [u8]) -> SResult {
```

Comment:
I'm not sure I agree with this. `encode` keeps the code uniform with respect to the ieee802154 code, which uses the same set of macros I used for header encoding/decoding. The behavior when the buffer is not big enough is also already specified by the `SResult` type. There is also no real set length for TLVs because the set of TLVs vary wildly depending on which MLE message is being sent. In other words, the responsibility for determining how much space to allocate falls on whoever is making the MLE message.